### DISCUSSION TITLE
Many uv_write callbacks delayed?  When do they actually get called?

### DISCUSSION THREAD

**Discussion by jkcko** [libuv/libuv#4033] [Q&A] [answered]
I have a libuv program running on an Nvidia Jetson Linux SBC which reads data from a High Speed USB interface and creates data records which are written to the SSD.  The reading from the USB is done in a thread which will block on the libusb bulk transfer to read the data.
The data from USB is written to a buffer pool with read/write/done indexes.
 - read index : updated in USB read thread where the USB data read is last stored up to before a call to uv_async_send()
 - write index : updated in the async callback for the uv_loop when it has called uv_write for the data
 - done index : updated in the uv_write callback

What I'm finding is that I'm occasionally running out of buffers in the pool when the done index hasn't incremented for some time (i.e. uv_write callbacks not being called in a timely manner).  Note that the uv_write is to a uv_pipe opened on a uv_file returned by uv_fs_open().

The pool size is 32 entries and the data may be up to ~2k bytes each.  The input data would max out at ~300kbps.
Surely the SSD writes can keep up with the input data right so why would the uv_write callbacks be delayed by so much and is there anything I can do in libuv to make sure the writes are being processed in a timely manner or push it along if the buffer pool is empty or low?

I can increase the buffer pool but that may only mask the problem temporarily without understanding why it so many uv_writes can be backed up.

Does the uv_write callback get called when the OS write is done or when the the disk is flushed.  I would think the former as disk flushing is managed by the OS.

**Comment by vtjnash** [comment]
`uv_pipe` does not support `uv_file` objects very well (mostly because epoll does a poor job with them), because, well, it is for `pipe` objects. I would have thought you would require ec8abc16af867b3eb9ffc6e3ddff4d8dfd710a55 to get any reasonable sort of behavior here. For `uv_file`, the interface is `uv_fs_write`.

**Reply by jkcko** [reply]
Is there another way to wrap a file into a stream object?  In this particular case I am outputting to a file but the output destination could also be a TCP socket or real pipe.

Since I'm only writing on the pipe object what role does epoll play here?  Is it required for the write request callback?

**Reply by jkcko** [reply]
After more study of the documents and the source code there isn't any other way to deal with files like other streams.
uv_files are not a type of uv_handle_t and uv_walk will not iterate over open file handles.

As for my questions the uv_fs_write callback doesn't mean the file is actually flushed to disk.  Which is the expected behaviour.

So it looks like uv_file handling MUST always be handled separately from any other stream type.

**Reply by bnoordhuis** [reply, marked as answer]
> So it looks like uv_file handling MUST always be handled separately from any other stream type.

Correct.

In the past we've discussed generalizing the streams API to support things like files or TLS but in the case of files that would be an opaque wrapper around uv_fs_write() and friends.

That kind of "API sugar" (or maybe I should say "obscuring") goes against libuv's design philosophy.
